respy/conditional_draws.py

```diff
@@ -54,95 +55,128 @@ def create_draws_and_log_prob_wages(
+        sigma_squared = cov[choice, choice] + meas_sds[choice] ** 2
+        sigma = np.sqrt(sigma_squared)
+        for i in range(dim):
+            updated_mean[i] = cov[choice, i] * shock / sigma_squared
+        loglike[0] = invariant - np.log(sigma) - shock ** 2 / (2 * sigma_squared)
+    else:
+        for i in range(dim):
+            updated_mean[i] = 0
+        loglike[0] = 0
 
-    ``extended_cholcov_t`` and ``state`` are changed in-place for performance reasons.
 
-    The function can handle missings in the measurements.
+def update_cholcov_with_measurement_error(shocks_cholesky, meas_sds, n_wages):
+    """Make a Kalman covariance updated for all possible cases.
 
     Parameters
     ----------
-    state : numpy.ndarray
-        Array with shape (n_choices,) containing initial state vectors.
-    measurement : float
-        The measurement that is incorporated through the Kalman update.
-    extended_cholcov_t : numpy.ndarray
-        Array with shape (n_states + 1, n_states + 1) that contains the transpose of the
-        Cholesky factor of the state covariance matrix in the lower right block and
-        zeros everywhere else.
-    meas_sd : float
-        The standard deviation of the measurement error.
-    choice : numpy.uint16
-        Observed choice. Determines on which element of the state vector is measured
-        by the measurement.
+    shocks_cholesky : numpy.ndarray
+        cholesky factor of the covariance matrix before updating. Has
+        dimension (n_choices, n_choices)
+
+    meas_sds: numpy.ndarray
+        the standard deviations of the measurement errors. Has length n_wages.
+
+    n_wages : int
+        number of wage sectors.
 
     Returns
     -------
-    log_prob_wage : float
-        The log likelihood of the observed wage.
+    updated_chols : numpy.ndarray
+        Array of (shape n_wages + 1, n_choices, n_choices) with the cholesky factors
+        of the updated covariance matrices for each possible observed shock. The
+        last element corresponds to not observing any shock.
+
+
+    We use a square-root implementation of the Kalman filter to avoid taking any
+    Cholesky decompositions which could fail due to numerical error.
 
     References
     ----------
     Robert Grover Brown. Introduction to Random Signals and Applied Kalman Filtering.
         Wiley and sons, 2012.
 
     """
-    # Extract dimensions.
-    m = len(extended_cholcov_t)
-    nfac = m - 1
+    n_choices = len(shocks_cholesky)
 
-    # This is the invariant part of a normal probability density function.
-    invariant = np.log(1 / (2 * np.pi) ** 0.5)
+    updated_chols = np.zeros((n_wages + 1, n_choices, n_choices))
 
-    # Skip missing wages.
-    if np.isfinite(measurement):
-        # Construct helper matrix for qr magic.
-        extended_cholcov_t[0, 0] = meas_sd
-        for f in range(1, m):
-            extended_cholcov_t[f, 0] = extended_cholcov_t[f, choice + 1]
+    for choice in range(n_wages):
```

Comment:
Above you used ``i`` while looping over ``n_choices`` because ``choice`` was already taken. As both functions are so close together, maybe also ``i``?

Reply:
This will be deleted anyways when I switch to the numpy qr.